ledger/participant-integration-api/src/main/scala/platform/store/dao/BufferedStreamsReader.scala

```diff
@@ -0,0 +1,109 @@
+  * @param metrics Daml metrics.
+  * @param name The name of the Ledger API stream. Used as a discriminator in metric registry names construction.
+  * @param executionContext The execution context
+  * @tparam PERSISTENCE_FETCH_FILTER The Ledger API streams filter type of fetches from persistence.
+  * @tparam API_RESPONSE The API stream response type.
+  */
+class BufferedStreamsReader[PERSISTENCE_FETCH_FILTER, API_RESPONSE](
+    inMemoryFanoutBuffer: EventsBuffer[TransactionLogUpdate],
+    persistenceFetch: PersistenceFetch[PERSISTENCE_FETCH_FILTER, API_RESPONSE],
+    eventProcessingParallelism: Int,
+    metrics: Metrics,
+    name: String,
+)(implicit executionContext: ExecutionContext) {
+  private val bufferReaderMetrics = metrics.daml.services.index.BufferedReader(name)
+
+  /** Serves processed and filtered events from the buffer, with fallback to persistence fetches
+    * if the bounds are not within the inclusive buffer range bounds.
+    *
+    * @param startExclusive The start exclusive offset of the search range.
+    * @param endInclusive The end inclusive offset of the search range.
+    * @param persistenceFetchFilter The filter used for fetching the Ledger API stream responses from persistence.
+    * @param bufferSliceFilter The filter used for filtering when searching within the buffer.
+    * @param toApiResponse To Ledger API stream response converter.
+    * @param loggingContext The logging context.
+    * @tparam BUFFER_OUT The output type of elements retrieved from the buffer.
+    * @return The Ledger API stream source.
+    */
+  def streamUsingBuffered[BUFFER_OUT](
+      startExclusive: Offset,
+      endInclusive: Offset,
+      persistenceFetchFilter: PERSISTENCE_FETCH_FILTER,
+      bufferSliceFilter: TransactionLogUpdate => Option[BUFFER_OUT],
+      toApiResponse: BUFFER_OUT => Future[API_RESPONSE],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Source[(Offset, API_RESPONSE), NotUsed] = {
+    def bufferSource(
+        bufferSlice: Vector[(Offset, BUFFER_OUT)]
+    ): Source[(Offset, API_RESPONSE), NotUsed] =
+      if (bufferSlice.isEmpty) Source.empty
+      else
+        Source(bufferSlice)
+          .mapAsync(eventProcessingParallelism) { case (offset, payload) =>
+            bufferReaderMetrics.fetchedBuffered.inc()
+            Timed.future(
+              bufferReaderMetrics.conversion,
+              toApiResponse(payload).map(offset -> _)(ExecutionContext.parasitic),
+            )
+          }
+
+    val source = Source
+      .unfoldAsync(startExclusive) {
+        case scannedToInclusive if scannedToInclusive < endInclusive =>
+          Future {
+            inMemoryFanoutBuffer.slice(scannedToInclusive, endInclusive, bufferSliceFilter) match {
+              case BufferSlice.Inclusive(slice) =>
+                val sourceFromBuffer = bufferSource(slice)
+                val nextChunkStartExclusive = slice.lastOption.map(_._1).getOrElse(endInclusive)
```

Comment:
Why not always use endInclusive

Reply:
Because IMFO has a max slice size parameter whereby it won't return matching items up to endInclusive if the startExclusive - endInclusive range is too large